### PR TITLE
Enable event handlers to indicate chart should be re-drawn

### LINKED ISSUE #593: Enable event handlers to indicate chart should be re-drawn
Chart.js allows plugins to set `args.changed`, so the chart can re-render automatically after processing the event completely. This is important when there are many plugins that require re-draws. It gets slow and messy if every plugin does the draw (or update!) in their event handlers.

So I think these handles should be allowed to set/return a flag indicating a redraw should be done.

_Originally posted by @kurkle in https://github.com/chartjs/chartjs-plugin-annotation/issues/589#issuecomment-1002426614_

### PR DESCRIPTION
**Pull request by stockiNail**
Fix #593

### CODE REVIEW

File: src/events.js [hunk 1]
```diff
@@ -72,26 +70,29 @@ function handleMoveEvents(state, event, options) {
...
-  dispatchMoveEvents(context, 'enter', elements, previous);
+  return dispatchMoveEvents(context, 'leave', previous, elements) || dispatchMoveEvents(context, 'enter', elements, previous);
 }
 
 function dispatchMoveEvents({state, event}, hook, elements, checkElements) {
+  let changed;
   for (const element of elements) {
     if (checkElements.indexOf(element) < 0) {
-      dispatchEvent(element.options[hook] || state.listeners[hook], element, event);
+      changed = changed || dispatchEvent(element.options[hook] || state.listeners[hook], element, event);
```
Comment (kurkle) [inline src/events.js:80]: here too, dispatchEvent will not be called anymore when changed is true.

File: src/annotation.js [hunk 2]
```diff
@@ -101,7 +101,7 @@ export default {
 
   beforeEvent(chart, args, options) {
     const state = chartStates.get(chart);
-    handleEvent(state, args.event, options);
+    args.changed = handleEvent(state, args.event, options) === true;
```
[the remark this answers is not in the document]
Comment (stockiNail) [reply src/annotation.js:104]: yes, smoking I was thinking about that! ;)

File: src/events.js [hunk 3]
```diff
@@ -72,26 +70,29 @@ function handleMoveEvents(state, event, options) {
   state.hovered = elements;
 
   const context = {state, event};
-  dispatchMoveEvents(context, 'leave', previous, elements);
-  dispatchMoveEvents(context, 'enter', elements, previous);
+  return dispatchMoveEvents(context, 'leave', previous, elements) || dispatchMoveEvents(context, 'enter', elements, previous);
```
[the remark this answers is not in the document]
Comment (stockiNail) [reply src/events.js:73]: YES... that's why I had a strange behavior on line annotation sample.

File: src/annotation.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (stockiNail) [reply src/annotation.js]: should be ok  now

File: src/events.js (hunk 1 above)
Comment (stockiNail) [reply src/events.js]: should be ok  now

File: src/events.js [hunk 4]
```diff
@@ -72,26 +70,29 @@ function handleMoveEvents(state, event, options) {
...
+  return changed;
 }
 
 function handleClickEvents(state, event, options) {
   const listeners = state.listeners;
   const elements = getElements(state, event, options.interaction);
+  let changed;
   for (const element of elements) {
-    dispatchEvent(element.options.click || listeners.click, element, event);
+    changed = changed || dispatchEvent(element.options.click || listeners.click, element, event);
```
[the remark this answers is not in the document]
Comment (stockiNail) [reply src/events.js]: should be ok  now

File: src/annotation.js [hunk 5]
```diff
@@ -101,7 +101,7 @@ export default {
 
   beforeEvent(chart, args, options) {
     const state = chartStates.get(chart);
-    handleEvent(state, args.event, options);
+    args.changed = handleEvent(state, args.event, options);
```
Comment (kurkle) [inline src/annotation.js:104]:
```suggestion
    if (handleEvent(state, args.event, options)) {
        args.changed = true;
    }
```
Should not set the `changed` to false if some other plugin already set it to true before the event was passed to this plugin.

**Review by kurkle — DISMISSED** [review]

**Review by kurkle — APPROVED** [review]

**Pull request merged**